test/regtest.py

```diff
@@ -104,43 +114,149 @@ def metadata(file):
-          print red('FAILED '),
-          failed += 1
+          str_result += 'FAILED '
 
-        print '  [%.2fs]' % round(elapsed, 2)
-        sys.stdout.flush()
-  
-except KeyboardInterrupt:
-  pass
+        str_result += '  [%.2fs]' % round(elapsed, 2)
 
-print
-print ' PASSED count:', passed
-print ' FAILED count:', failed
+        # log the info for this test
+        logging.info(str_result)
 
-if timeouts > 0:
-  print 'TIMEOUT count:', timeouts
-
-if unknowns > 0:
-  print 'UNKNOWN count:', unknowns
+        # returns a status code based on the results of the test
+        if passed:
+            return PASSED
+        elif timedout:
+            return TIMEDOUT
+        elif unknown:
+            return UNKNOWN
+        else:
+            return FAILED
+
+
+def init_worker():
+    """
+    Set up the worker processes to ignore SIGINT altogether. This confines all
+    the cleanup code to the parent process.
+
+    Reference: http://stackoverflow.com/a/6191991/3342427
+    """
+    #signal.signal(signal.SIGINT, signal.SIG_IGN)
+
+def main():
+    """
+    Main entry point for the test suite.
+    """
+    t0 = time.time()
+    num_cpus = multiprocessing.cpu_count()
+
+    # configure the CLI
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--exhaustive", help="be exhaustive", action="store_true")
+    parser.add_argument("--threads", action="store", dest="n_threads", default=num_cpus, type=int,
+                        help="execute regressions using the selected number of threads in parallel")
+    parser.add_argument("--log", action="store", dest="log_level", default="DEBUG", type=str,
+                        help="sets the logging level (DEBUG, INFO, WARNING)")
+    parser.add_argument("--output_log", action="store", dest="log_path", type=str,
+                        help="sets the output log path. (std out by default)")
+    args = parser.parse_args()
+
+
+    # configure the logging
+    log_format = ''
+    log_level = logging.DEBUG
+
+    # add more log levels later (if needed)
+    if args.log_level.upper() == "INFO":
+        log_level = logging.INFO
+    elif args.log_level.upper() == "WARNING":
+        log_level = logging.WARNING
+
+    # if the user supplied a log path, write the logs to that file.
+    # otherwise, write the logs to std out.
+    if args.log_path:
+        logging.basicConfig(filename=args.log_path, format=log_format, level=log_level)
+    else:
+        logging.basicConfig(format=log_format, level=log_level)
+
+    logging.debug("Creating Pool with '%d' Workers" % args.n_threads)
+    p = ThreadPool(processes=args.n_threads, initializer=init_worker)
+
+    # start the tests
+    logging.info("Running regression tests...")
+    passed = failed = timeouts = unknowns = 0
+
+    # start processing the tests. this blocks the main thread
+    async_results = []
+    for test in glob.glob("./**/*.c"):
+        async_results.append(p.apply_async(process_test, args=(test, args,)))
```

Comment:
The logic here seems a bit flawed, since only one task is being created per file, with a single return value. When the exhaustive flag is passed, then each file has several tests, corresponding to different memory models and verifiers — i.e., given in `meta['memory']` and `meta['verifiers']`.

The async task creation loop should thus look more like this:
````
for test in glob.glob(…):
  for memory in meta['memory'][…]:
    for verifier in meta['verifier'][…]:
      …
      p.apply_async(process_test, …)
````
Otherwise, the `--exhaustive` flag doesn’t really do anything.